### PR TITLE
refactor (jkube-kit) : Move Dockerfile opinionated ImageConfig logic from `ConfigHelper.initImageConfiguration` to a dedicated generator

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

+ ConfigHelper modifies the provided set of ImageConfigurations when it detects a `Dockerfile` in project root directory. This is actually a feature called Simple Dockerfile Mode. However, this mutation of ImageConfiguration from within `initImageConfiguration` does not look appropriate. It can be moved to a dedicated generator that will get activated when `Dockerfile` is present in project base directory.
+ Add a new generator SimpleDockerfileGenerator, it adds a simple ImageConfiguration with Dockerfile configured in build and adds it to existing set of ImageConfigurations.
+ Remove check from BaseGenerator to skip generation if it's simple Dockerfile mode. Move it to SimpleDockerfileGenerator.isApplicable instead

We also need to adjust the documentation for adding this new generator. I'll do that in a follow up PR.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [X] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [X] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [X] I have performed a self-review of my code
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I have added tests that prove my fix is effective or that my feature works
 - [X] New and existing unit tests pass locally with my changes
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
I recall we might have discussed this a few weeks ago.
We need to ensure we have enough tests that guarantee that in case the order of generators is altered it gets reported.
This idea is really good to clean out the code but it has the fragility of depending on the generator order.

**Comment by rohanKanojia** [comment]
@manusa : I had added a test in SimpleDockerFileGeneratorTest to verify that generator doesn't modify existing ImageConfiguration when they have been altered by any other generator.

Shall I add an additional test to assert the order of generators in the default profile?

**Comment by rohanKanojia** [comment]
I've added tests verifying the generators execution order in the default profile in b5d9de6753534bc9b94f663d5c3941e83ac1b188 . Please check if it's in line with your expectations.

File: jkube-kit/generator/dockerfile-simple/src/main/java/org/eclipse/jkube/generator/dockerfile/simple/SimpleDockerfileGenerator.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import static org.eclipse.jkube.kit.common.util.PropertiesUtil.getValueFromProperties;
+
+public class SimpleDockerfileGenerator extends BaseGenerator {
+  public SimpleDockerfileGenerator(GeneratorContext context) {
+    super(context, "dockerfile-simple");
+  }
+
+  @Override
+  public boolean isApplicable(List<ImageConfiguration> configs) {
+    return shouldAddGeneratedImageConfiguration(configs) &&
```
[the remark this answers is not in the document]
Comment (rohanKanojia) [reply jkube-kit/generator/dockerfile-simple/src/main/java/org/eclipse/jkube/generator/dockerfile/simple/SimpleDockerfileGenerator.java:38]: You're right. Thanks a lot for noticing. Let me update it.

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTaskTest.java [hunk 2]
```diff
@@ -137,4 +159,43 @@ public Property<Boolean> getSkipBuild() {
...
+        .properties(new Properties()).build();
+    when(taskEnvironment.project.getExtensions().getByType(KubernetesExtension.class)).thenReturn(extension);
+    final KubernetesBuildTask buildTask = new KubernetesBuildTask(KubernetesExtension.class);
+    GeneratorContext generatorContext = buildTask.initGeneratorContextBuilder().build();
+
+    // When
+    List<Generator> generatorList = createUsableGeneratorList(generatorContext);
+
+    // Then
+    Assertions.assertThat(generatorList)
```
Comment (sunix) [inline gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTaskTest.java:186]: for readability purpose, could you make Assertions.assertThat statically imported ?

File: jkube-kit/generator/dockerfile-simple/pom.xml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>org.eclipse.jkube</groupId>
+      <artifactId>jkube-kit-generator-api</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
```
Comment (sunix) [inline jkube-kit/generator/dockerfile-simple/pom.xml:40]: I would keep the scope declaration consistent:
- either you don't declare them and assume it is coming from the dependencyManagement
- either you declare them everywhere (also for org.assertj)

File: kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java [hunk 4]
```diff
@@ -84,6 +84,7 @@ void setUp(@TempDir Path temporaryFolder) throws Exception {
     when(mavenProject.getCompileClasspathElements()).thenReturn(Collections.emptyList());
     when(mavenProject.getDependencies()).thenReturn(Collections.emptyList());
     when(mavenProject.getDevelopers()).thenReturn(Collections.emptyList());
+    when(mavenProject.getBasedir()).thenReturn(temporaryFolder.toFile());
```
Comment (sunix) [inline kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java:87]: is this related to refactor ?

Comment (rohanKanojia) [reply kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/develop/WatchMojoTest.java:87]: `SimpleDockerfileGenerator.isApplicable` method requires project base directory to be provided. `mavenProject` is mocked with deep stubs; when we invoke `mavenProject.getBasedir()` it is returning some mocked object which is causing NPE when this base directory is checked in  `SimpleDockerfileGenerator.isApplicable`

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTaskTest.java [hunk 5]
```diff
@@ -137,4 +159,43 @@ public Property<Boolean> getSkipBuild() {
...
+    final KubernetesBuildTask buildTask = new KubernetesBuildTask(KubernetesExtension.class);
+    GeneratorContext generatorContext = buildTask.initGeneratorContextBuilder().build();
+
+    // When
+    List<Generator> generatorList = createUsableGeneratorList(generatorContext);
+
+    // Then
+    assertThat(generatorList)
+        .hasSize(12)
+        .satisfies(g -> Assertions.assertThat(g.get(0)).isInstanceOf(SimpleDockerfileGenerator.class))
```
Comment (sunix) [inline gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTaskTest.java:188]: Same here, would be nice to have static imports for Assertions.assertThat

**Review by sunix — APPROVED** [review]

File: gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTaskTest.java (hunk 5 above)
Comment (rohanKanojia) [reply gradle-plugin/kubernetes/src/test/java/org/eclipse/jkube/gradle/plugin/task/KubernetesBuildTaskTest.java:188]: Sorry, let me fix it.

**Comment by manusa** [comment]
> I've added tests verifying the generators execution order in the default profile in [b5d9de6](https://github.com/eclipse/jkube/commit/b5d9de6753534bc9b94f663d5c3941e83ac1b188) . Please check if it's in line with your expectations.

This works, however is extremely brittle and hard to maintain (which probably defeats the purpose). The idea is to create a test that ensures the behavior (not sure how since I haven't dug into the code), test should fail in the event of a modification in the generator order (default profile).

**Comment by rohanKanojia** [comment]
@manusa There is a `log.verbose` statement in DefaultGeneratorManager that prints the generators being invoked: 

https://github.com/eclipse/jkube/blob/4cfb00728d455273827d0552376f304104fbed24/jkube-kit/generator/api/src/main/java/org/eclipse/jkube/generator/api/DefaultGeneratorManager.java#L54-L56

Shall I try adding log verifications for this instead?

**Comment by manusa** [comment]
> Shall I try adding log verifications for this instead?

If the check consists in checking the generator print order, then that's mostly the same of what we have now.
We either need to ensure that this generator gets printed the very first always, or that its behavior is consistent.
When I talk about behavior is having a test that would fail if the order of generators was altered (maybe we already have one like this, but in that case we should state explicitly or add a case where this is reported)

File: kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/BuildMojoTest.java [hunk 6]
```diff
@@ -58,4 +73,35 @@ void buildServiceConfigBuilder_shouldReturnNonNullResourceConfigIfConfigured() {
             .returns("200m", c -> c.getLimits().get("cpu"))
             .returns("1Gi", c -> c.getRequests().get("memory"));
     }
+
+    @Test
+    @DisplayName("when Dockerfile present in project base directory, then generate opinionated ImageConfiguration for Dockerfile")
+    void simpleDockerfileModeWorksAsExpected(@TempDir File temporaryFolder) throws IOException {
```
Comment (manusa) [inline kubernetes-maven-plugin/plugin/src/test/java/org/eclipse/jkube/maven/plugin/mojo/build/BuildMojoTest.java:79]: These tests (task + Mojo) work, but they aren't exactly where they should actually be.
I'll try to give take it for a spin and see if we can make this easier.

**Review by manusa — APPROVED** [review]
LGTM, thx!
